clients/stellar-relay-lib/src/connection/connector/message_reader.rs

```diff
@@ -43,24 +43,33 @@ pub(crate) async fn poll_messages_from_stellar(
 		}
 
 		// check for messages from Stellar Node.
-		let xdr = match read_message_from_stellar(&mut connector).await {
-			Err(e) => {
+		// if reading took too much time, flag it as "disconnected"
+		let xdr = match timeout(
+			Duration::from_secs(READ_TIMEOUT_IN_SECS),
+			read_message_from_stellar(&mut connector)
+		).await {
+			Ok(Ok(xdr)) => xdr,
+			Ok(Err(e)) => {
 				error!("poll_messages_from_stellar(): {e:?}");
-				break;
+				break
 			},
-			Ok(xdr) => xdr,
+			Err(_) => {
+				error!("poll_messages_from_stellar(): timed out");
+				break
+			}
 		};
 
 		match connector.process_raw_message(xdr).await {
-			Ok(Some(stellar_msg)) =>
-			// push message to user
-			{
-				if let Err(e) = send_to_user_sender.send(stellar_msg.clone()).await {
+			Ok(Some(stellar_msg)) => {
+				// push message to user
+				let stellar_msg_as_base64_xdr = stellar_msg.to_base64_xdr();
+				if let Err(e) = send_to_user_sender.send(stellar_msg).await {
 					warn!("poll_messages_from_stellar(): Error occurred during sending message {} to user: {e:?}",
-						String::from_utf8(stellar_msg.to_base64_xdr())
-						.unwrap_or_else(|_| format!("{:?}", stellar_msg.to_base64_xdr()))
-					);
+					String::from_utf8(stellar_msg_as_base64_xdr.clone())
+					.unwrap_or_else(|_| format!("{stellar_msg_as_base64_xdr:?}"))
+				);
 				}
+				tokio::task::yield_now().await;
```

Comment:
For the `break`s, I wanted a message to be sent back to the user that the stellar-relay will stop functioning.
Breaking out of the loop will stop the thread entirely. 

I chose to yield _only_ when `message_reader()` is doing fine. This will hit _pause_ so that other tasks will run. Then if its turn has come, the loop will go on.

Reply:
Makes sense, you want to make sure the `StellarOverlayConnection` process the message before continuing. This is quite complex to model in my mind since if we don't yield, then we probably continue with the next iteration, and at some point something when we hit `.await()` it will yield. But then perhaps `send_to_user_sender.is_closed()` is `true` and we break without the chance for the receiver to do something with the message.

So it "may" be redundant, but it doesn't hurt at all. Did you notice that without this yield something was failing?

Reply:
I agree it makes sense, thanks for explaining @b-yap.

Reply:
>  Did you notice that without this yield something was failing?

@gianfra-t So I yielded because in the tokio-console, this thread almost always was **BUSY**, giving no turn for other tasks to run. It was always trying to read messages from Stellar Relay. A yield for a few seconds wouldn't hurt.

> So it "may" be redundant,

True, it could be. But it is a loop, no one is awaiting `fn poll_messages_from_stellar()`.

Reply:
> no one is awaiting fn poll_messages_from_stellar()

True! I meant that inside the loop itself, on the other operations like `read_message_from_stellar` we await so unless that one is also never yielding, the overall process will (I think?). But it is more explicit this way that we want to yield at this point so it's fine by me 👍.